### PR TITLE
Sensitivity-Specificity-AUC section

### PR DESCRIPTION
**Pull request by ananthagarwal**

### CODE REVIEW

**Comment by ananthagarwal** [comment]
This is my first draft of the precision recall AUC section. I plan on working on the wording over the next day or two, but please review my example and concepts covered so that I know I'm on the right track.

File: notebooks/ch16/classification_precision_recall.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+```
+
+    Test set accuracy: 0.944
+
+
+Using just the occurrence of four words, we have built a model that accurately classifies about 94% of the emails in the test set, which at first glance is a good result. However, if we take a closer look at our data we realize that our classifier isn't performing as well as we thought. 
+
+The accuracy we computed is the fraction of predictions that the model correctly classified. The fact that the accuracy is high implies that most of the model's classifications are true positives and true negatives, which occur when the model correctly labels the positive class (spam) and negative class (ham) respectively. On the other hand, there are two types of errors a model can make:
+
+* False positive: a ham email gets flagged as spam and filtered out of the inbox. Generally, a false positive occurs when the model assigns a data point to the positive class (spam) whereas it actually belonged in the negative class (ham).
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: This says "Generally", but goes on to reference the specific example in parentheses. I think it'd be clearer to first write the general definition of FP, then use the example to reinforce the concept. Likewise for the FN definition/example in the following line.

File: notebooks/ch16/classification_precision_recall.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+* False negative:  a spam email gets mislabelled as ham and ends up in the inbox. Generally, a false negative occurs when the model assigns a data point to the negative class (ham) whereas it actually belonged in the positive class (spam).
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: (nit) Supposedly, using one "l" in "mislabeled" is the "American" way of writing it.

File: notebooks/ch16/classification_precision_recall.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+
+![png](classification_precision_recall_files/classification_precision_recall_11_1.png)
+
+
+See [here](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html#sklearn.metrics.confusion_matrix) for more on scikit-learn's confusion matrix generator.
+
+Adding together the number of true positives and false negatives, we know the training dataset contains 36 spam emails, but the model only detected 3 of them. Despite its high accuracy, the model identified $\frac{1}{12}$ of spam emails, which is quite a poor performance. 
+
+What if we used a classifier `ham_only` that simply labelled every email as ham? We can calculate its accuracy simply by knowing the number of ham emails in the dataset: $$ \text{Accuracy} = \frac{TP + TN}{TP + FP + TN + FN} = \frac{0 + 714}{0 + 0 + 714 + 36} = .95$$
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: (nit again) labeled/labelled

File: notebooks/ch16/classification_precision_recall.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+
+
+This model, which is essentially useless as a classifier, surprisingly achieves a better accuracy than our EDA-driven model. Clearly accuracy alone can be a misleading indicator of model performance. This is especially characteristic of **class-imbalanced datasets**, in which a vast majority of labels belong in one class over the other. In this case, most of our emails are ham. The confusion matrix shows that the model based on the occurrences of three words had many more false negatives than true positives, which implies that it didn't have enough features to reliably identify spam, but because the dataset is predominantly ham, it had a deceptively high accuracy. Another common example of class imbalance is disease detection when the frequency of the disease in a population is low.
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: - It might be worth clarifying out why `ham_only` is a useless classifier; in all the terminology and new concepts floating around, readers might just get lost in the words and it's useful to provide a bigger picture.
- "model based on the occurrences of three words" should be "four words" - please, click, money, business. I also think that sentence is a bit of a run-on; might be worth rewording that so it flows better?

File: notebooks/ch16/classification_precision_recall.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+
+We now turn to precision and recall, which are better suited metrics for evaluating class-imbalanced datasets.
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: > We now turn to precision and recall, two metrics that are better suited for evaluating class-imbalanced datasets.

File: notebooks/ch16/classification_precision_recall.md [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+
+### Precision
+
+**Precision** (also called **positive predictive value**) measures the proportion of positively classified data that are actually positive.
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: Hmm, I think in class we referred to it more often as "specificity" instead of "positive predictive value"

File: notebooks/ch16/classification_precision_recall.md [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+We now turn to precision and recall, which are better suited metrics for evaluating class-imbalanced datasets.
+
+### Precision
+
+**Precision** (also called **positive predictive value**) measures the proportion of positively classified data that are actually positive. 
+
+$$ \text{Precision} = \frac{TP}{TP + FP} $$ 
+
+A model with high precision accurately classifies data that truly has positive labels, whereas a model with low precision incorrectly picks out aspects of the positive class in truly negative data with high frequency (many false positives).
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: "incorrectly picks out aspects of the positive class in truly negative data" was a little confusing for me. It's theoretically sound - since it incorrectly picks up trends that it thinks is "positive" in truly negative data - but I think for readers, it's better to just define what the classifier's actually doing:

> whereas a model with low precision incorrectly labels truly negative data as positive (i.e. false positive).

File: notebooks/ch16/classification_precision_recall.md [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+
+The **classification threshold** is a value that determines what class a data point is assigned to; points on opposite sides of the threshold are labelled with different classes. Recall that logistic regression outputs a probability that the data point belongs to the positive class. If the probability output on $x_1$ is .90 and the output on $x_2$ is .1, then we can be fairly confident that $x_1$ is spam and $x_2$ is ham. However, if the probability associated with $x_3$ is around .55, then we have less confidence and based on the classification threshold, $x_3$ may be labelled as spam in some cases and ham in others. Depending on the problem statement, false positives and false negatives may have different importance, so set the classification threshold based on which error type you want to minimize. For email classification, false positives result in important emails being filtered out, so they are worse than false negative, in which a spam email winds up in the inbox. In medical settings, however, false negatives in a diagnostic test can be much more consequential than false positives.
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: - "they are worse than false negative" --> "they are worse than false negatives"
- While talking about whether we should prioritize FN/FP minimization, it's probably also a good idea to mention whether that corresponds to Precision or Recall
- labelled/labeled

File: notebooks/ch16/classification_precision_recall.md [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+
+
+
+
+![png](classification_precision_recall_files/classification_precision_recall_17_1.png)
+
+
+See [here](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html#sklearn.metrics.precision_recall_curve) for scikit-learn's precision-recall function.
+
+The **Area Under the Curve (AUC)** is a single number performance summary of the classifier. A large AUC means that the model has both high precision and recall, which implies that it has few false postivives and false negatives. The AUC for `words_list_model` is:
```
Comment (junseo-park) [inline notebooks/ch16/classification_precision_recall.md]: I figure you're not done with this portion yet; I think we should expand just a little bit about how the AUC is generated and what it represents

File: notebooks/ch16/classification_precision_recall.md (hunk 6 above)
Comment (ananthagarwal) [reply notebooks/ch16/classification_precision_recall.md]: Specificity is actually not the same as precision (specificity is the true negative rate).

File: notebooks/ch17/classification_precision_recall.md [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+
+
+
+
+![png](classification_precision_recall_files/classification_precision_recall_17_1.png)
+
+
+See [here](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html#sklearn.metrics.precision_recall_curve) for scikit-learn's precision-recall function.
+
+The **Area Under the Curve (AUC)** is a single number performance summary of the classifier. A large AUC means that the model has both high precision and recall, which implies that it has few false postivives and false negatives. The AUC for `words_list_model` is:
```
Comment (SamLau95) [inline notebooks/ch17/classification_precision_recall.md]: We should add a sentence here explaining how to calculate the AUC, what an AUC of 1 means, and how a random classifier performs.

**Comment by ananthagarwal** [comment]
The second draft is contained in classification_sensitivity_specificity.ipynb

I know I haven't explained AUC very well: can you please help me with phrasing it to make sense? One other thing I'm not sure about is why `ham_only` and `spam_only` both have an AUC of 0.5. Perhaps I calculated it wrong, but if not I would appreciate an explanation. 

One more confusion I would like cleared up: one of the main points of the section is that adjusting the classification threshold affects the relative number of false positives/negatives, but I haven't found an easy way to do this in scikit-learn, which seems to always use 0.5 as the default for logistic regression. The ROC curve indicates good threshold values, but how can we actually implement them in scikit-learn? Also, does it make sense to encourage more EDA-driven feature selection in the Summary to improve a model (if there are a lot of false positives/negatives) in addition to playing around with the classification threshold?

File: notebooks/ch17/classification_precision_recall.md (hunk 10 above)
Comment (ananthagarwal) [reply notebooks/ch17/classification_precision_recall.md]: I was able to address all your comments except this one (please read my longer comment below).

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+```
+
+    words_list_model test set accuracy: 0.96
+    ham_only test set accuracy: 0.96
+    spam_only test set accuracy: 0.04
+
+
+Using `words_list_model` classifies 96% of the test set emails correctly. Although this accuracy appears high, `ham_only` achieves the same accuracy by uniformally labeling everything as ham. This is cause for concern because the data suggests we can do just as well without a spam filter at all.
+
+Model accuracy alone can be a misleading indicator of model performance, so we must investigate further. For each of our models, we can compare the number of erroneous classifications to the number of correct classifications by building a **confusion matrix**. A confusion matrix for a binary classifier is a two-by-two table that contains the model predictions on one axis and the actual labels on the other.
```
Comment (SamLau95) [inline notebooks/ch17/classification_sensitivity_specificity.md]: Let's add a hook at the beginning of this sentence to make the transition smoother:

> As the accuracies above show, model accuracy alone can be a misleading indicator of model performance. We can understand the model's predictions in greater depth using a **confusion matrix**. A confusion matrix for a binary classifier is a two-by-two heatmap that contains the model predictions on one axis and the actual labels on the other.

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+
+$$ \text{Sensitivity} = \frac{TP}{TP + FN} $$
+
+From our discussion of confusion matrices, you should recognize the expression $TP + FN$ as the sum of the entries in the first row. Building confusion matrices allows us to easily compare the sensitivities of our models:
+
+* `ham_only`: $\frac{0}{0 + 42} = 0$
+* `spam_only`: $\frac{42}{42 + 0} = 1$
+* `words_list_model`: $\frac{18}{18 + 24} \approx .429$
+
+Since `ham_only` has no true positives, it has the worst possible sensitivity value of 0. On the other hand, `spam_only` has an abysmally low accuracy but it has the best possible sensitivity value of 1. The low sensitivity of `words_list_model` indicates that it frequently fails to mark spam emails as such. Nonetheless, it outperforms `ham_only`, which will prove significant when we examine ROC curves later.
```
Comment (SamLau95) [inline notebooks/ch17/classification_sensitivity_specificity.md]: I don't think we need to mention "will prove significant when we examine ROC curves later." since ROC curves have no meaning for the reader right now and the importance of sensitivity will be clear when we get to the curve.

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+
+$$ \text{Precision} = \frac{TN}{TN + FP} $$ 
+
+We can use values from the confusion matrices to compare the specificities of our models:
+
+* `ham_only`: $\frac{758}{758 + 0} = 1$
+* `spam_only`: $\frac{0}{0 + 758} = 0$
+* `words_list_model`: $\frac{752}{752 + 6} \approx .992$
+
+As with sensitivity, the worst and best specificities are 0 and 1 respectively. Notice that `ham_only` has the best specificity and worst sensitivity, while `spam_only`has the worst specificity and best sensitivity. Since these models only predict one label, they will misclassify all instances of the other label, which is reflected in the extreme sensitivity and specificity values. This contrast is not nearly as pronounced for `words_list_model`.
```
Comment (SamLau95) [inline notebooks/ch17/classification_sensitivity_specificity.md]: nit: need space after `spam_only`.

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 14]
```diff
@@ -0,0 +1,486 @@
...
+
+We can use values from the confusion matrices to compare the specificities of our models:
+
+* `ham_only`: $\frac{758}{758 + 0} = 1$
+* `spam_only`: $\frac{0}{0 + 758} = 0$
+* `words_list_model`: $\frac{752}{752 + 6} \approx .992$
+
+As with sensitivity, the worst and best specificities are 0 and 1 respectively. Notice that `ham_only` has the best specificity and worst sensitivity, while `spam_only`has the worst specificity and best sensitivity. Since these models only predict one label, they will misclassify all instances of the other label, which is reflected in the extreme sensitivity and specificity values. This contrast is not nearly as pronounced for `words_list_model`. 
+
+Both sensitivity and specificity are needed to thoroughly evaluate a model, but it is not yet clear how we can use them to conclusively determine which model is better. Before we delve into this, we will take a slight detour to discussion classification threshold.
```
Comment (SamLau95) [inline notebooks/ch17/classification_sensitivity_specificity.md]: What do you think about this revision for the paragraph?

> Although sensitivity and specificity seem to describe different characteristics of a classifier, we draw an important connection between these two metrics using the classification threshold.

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+
+### Classification Threshold
+
+The **classification threshold** is a value that determines what class a data point is assigned to; points that fall on opposite sides of the threshold are labeled with different classes. Recall that logistic regression outputs a probability that the data point belongs to the positive class. If the probability output on $x_1$ is .90 and the output on $x_2$ is .1, then we can be fairly confident that $x_1$ is spam and $x_2$ is ham. However, if the probability associated with $x_3$ is around .55, there is greater uncertainty about the correct class. If the classification threshold is .50, $x_3$ is labeled as spam, but if the threshold is .60, $x_3$ is labeled as ham.
```
Comment (SamLau95) [inline notebooks/ch17/classification_sensitivity_specificity.md]: Can we just start with a quick example of using the threshold first, and then talk about uncertainty?

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 16]
```diff
@@ -0,0 +1,486 @@
...
+
+Although the goal is to maximize both sensitivity and specificity, they are inversely related so there is a tradeoff. Increasing sensitivity leads to a decrease in specificity and vice versa. For instance, reducing false negatives by lowering the classification threshold increases false positives on data that initially fell near the classification threshold but was correctly labeled with the negative class.
```
Comment (SamLau95) [inline notebooks/ch17/classification_sensitivity_specificity.md]: Instead of talking about sensitivity and specificity straight away, can you provide a short example of adjusting the classification threshold between ~3 values and showing how the confusion matrix changes? This will make the trade-off between sensitivity and specificity much clearer (since we "show, not tell") and provide better motivation for the ROC curve.

**Comment by SamLau95** [comment]
Also, you might want to look into https://seaborn.pydata.org/generated/seaborn.heatmap.html for heatmaps.

**Comment by ananthagarwal** [comment]
Quick clarification question: If I have a confusion matrix that indicates I have a lot of false positives and/or negatives, should my first course of action be 1. to adjust the classification threshold for the existing model or 2. start doing more EDA/feature selection to change the model itself in hopes of improving it?

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+
+
+Each row contains the body of an email and a `spam` column, which is `0` if the email is ham or `1` if it is spam.
+
+Let's compare the performance of three different classifiers:
+* `ham_only`: predicts 'ham' for every email.
+* `spam_only`: predicts 'spam' for every email.
+* `words_list_model`: predicts 'ham' or 'spam' based on the presence of certain words in the body of an email. 
+
+We construct `words_list_model` using the following procedure: Suppose we have a list of words `words_list` that we believe are common in spam emails: "please", "click", "money", "business", and "remove". We transform each email into a feature vector by setting the vector's $i$th entry to 1 if the $i$th word in `words_list` is contained in the email body and 0 if it isn't. For example, using our five chosen words and the email body "please remove by tomorrow", the feature vector would be $[1, 0, 0, 0, 1]$. This procedure generates the `1000 X 5` feature matrix $X$.
```
Comment (junseo-park) [inline notebooks/ch17/classification_sensitivity_specificity.md]: The first sentence has two colons in it; I think the first clause can be split off into its own sentence and ended with a period.

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+
+    Text(0.5,1,'words_list_model ROC Curve')
+
+
+
+
+![png](classification_sensitivity_specificity_files/classification_sensitivity_specificity_24_1.png)
+
+
+Notice that as we move from left to right across the curve, sensitivity increases and the specificity decreases. Generally, the best classification threshold corresponds to high sensitivity and specificity (low false positive rate). Picking a definitive value for the threshold depends on the classification problem and the relative costs of the two error types. Set the classification threshold based on which error type you want to minimize: minimizing false positives maximizes specificity and minimizing false negatives maximizes sensitivity. For email classification, false positives result in important emails being filtered out, so they are worse than false negatives, in which a spam email winds up in the inbox. In medical settings, however, false negatives in a diagnostic test can be much more consequential than false positives.
```
Comment (junseo-park) [inline notebooks/ch17/classification_sensitivity_specificity.md]: I think the last two sentences have a lot of commas that deter from their readability. Here's a suggestion? (I'm honestly not sure how to tackle this; it's wordy but every phrase does matter immensely. It could be that the current text is the best solution possible.)

> In email classification, our goal is to minimize false positives so that important emails do not get filtered out as spam. This is a tradeoff we are willing to make in exchange for higher specificity, in which spam emails may find themselves in the inbox due to a higher false negative rate. However, in a medical setting we prioritize minimizing specificity, since false negatives can be much more consequential than false positives.

File: notebooks/ch17/classification_sensitivity_specificity.md [hunk 19]
```diff
@@ -0,0 +1,486 @@
...
+
+
+
+    0.5
+
+
+
+### Summary
+
+To thoroughly evaluate a model, accuracy, sensitivity, specificity, and AUC ought to be considered together, especially for class-imbalanced datasets. Constructing a confusion matrix helps to visualize the ratio of correct predictions to errors. Based on these results, you can improve a model through further EDA-drive feature selection.
```
Comment (junseo-park) [inline notebooks/ch17/classification_sensitivity_specificity.md]: nit: "model, accuracy, sensitivity, specificity, and AUC" is an odd set of commas. We can probably reword this to avoid that:

> To thoroughly evaluate a model, we consider accuracy, sensitivity, specificity, and AUC. This is particularly important for class-imbalanced datasets, in which ...

**Comment by ananthagarwal** [comment]
Revamped ROC and AUC sections. Will continue to make wording improvements here and there.

**Comment by SamLau95** [comment]
Looks good to me! Merge away :)

**Comment by SamLau95** [comment]
Oh, and make sure you update the SUMMARY.md file before merging.
